Not getting success callback/ receipt with walletconnect with web3.js 1.3.3 - Closes #3891

nazarhussain opened the pull request:
## Description

We were using an assumption that if any provider contains `on` handler, it must also support `newBlockHeaders` event. Which was not the case for all providers and was not enforced by [EIP-1193](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1193.md). So for the flow of waiting for the receipt we reverted back only to polling technique. 

Fixes: #3891 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist:

- [x] I have selected the correct base branch.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My changes generate no new warnings.
- [x] I have tested my code on the live network.
- [x] I have checked the Deploy Preview and it looks correct.
- [x] I have updated the `CHANGELOG.md` file in the root folder.

nazarhussain committed:
:bug: Fix transaction receipt polling technique

nazarhussain committed:
:white_check_mark: Add unit test for the delay scenario

nazarhussain commented:
To come up with a solution which impact least on bandwidth usage and to providers which support `newBlockHeaders` I came up with this solution. 

1. If provider does not support `on` handler use polling anyway. 
2. If provider support `on` handler subscribe to `newBlockHeaders` event
3. Check after `blockHeaderTimeout` default to 10s, if block header event received or not 
4. If not, that's probably implies provider does not support that event and we fallback to polling. 

This approach will introduce no behaviour changes for providers which support `newBlockHeaders` events, and would introduce some wait time for others. 

@jdevcs  @spacesailor24 Let me know if that solutions seems feasible to you in reference to the comment posted on issue earlier https://github.com/ChainSafe/web3.js/issues/3891#issuecomment-916387467. If its ok to you guys then I will add few more tests as well. Also let me know if that change required to update some other places as well.

spacesailor24 commented on packages/web3-core-method/src/index.js:
```diff
@@ -547,22 +548,35 @@ Method.prototype._confirmTransaction = function (defer, result, payload) {
+        _ethereumCall.subscribe('newBlockHeaders', function (err, blockHeader, sub) {
+            blockHeaderArrived = true; 
+
+            if (err || !blockHeader) {
```
I understand this is the original code, but kinda weird to revert to polling if the subscription returns an error, no?

Spacesai1or commented on test/method.buildCall.js:
```diff
@@ -1117,6 +1117,67 @@ describe('lib/web3/method', function () {
             })
         });
+
+        it('should fallback to polling if provider support `on` but `newBlockHeaders` does not arrive in `blockHeaderTimeout` seconds', function (done) {
```
This doesn't test if `blockHeaderTimeout` works as expected if a response for `newBlockHeaders` isn't returned, correct?

nazarhussain replied:
Based on the sequence of assertions provider will expect `eth_getTransactionReceipt` call from the code, that can only happen when `blockHeaderTimeout` triggers and fallback to polling..

Spacesai1or commented on packages/web3-core-method/src/index.js:
```diff
@@ -547,22 +548,35 @@ Method.prototype._confirmTransaction = function (defer, result, payload) {
+        _ethereumCall.subscribe('newBlockHeaders', function (err, blockHeader, sub) {
+            blockHeaderArrived = true; 
+
+            if (err || !blockHeader) {
```
If the provider doesn't support `newBlockHeaders`, does the callback never get called? I'm confused when `!blockHeader` would occur

nazarhussain replied:
If provider does not support `newBlockHeaders` then this callback never triggered. The check here is a safety check when `newBlockHeaders` event fired but with some error.

Spacesai1or commented on packages/web3-core-method/src/index.js:
```diff
@@ -56,6 +56,7 @@ var Method = function Method(options) {
     this.transactionBlockTimeout = options.transactionBlockTimeout || 50;
     this.transactionConfirmationBlocks = options.transactionConfirmationBlocks || 24;
     this.transactionPollingTimeout = options.transactionPollingTimeout || 750;
+    this.blockHeaderTimeout = options.blockHeaderTimeout || 10; // 10 seconds
```
Any issue with cutting this down to 5 seconds?

nazarhussain replied:
Personally I was not sure about this default value. I will update it to 5.

Spacesai1or reviewed (changes requested):
Couple questions, but otherwise changes look good

Could use at least one other test if my test related question is correct, also needs `CHANGELOG.md` update

nazarhussain committed:
Merge branch '1.x' into nh/3891-receipt-callback

nazarhussain committed:
:wrench: Add configuration option "blockHeaderTimeout"

nazarhussain committed:
:pencil: Add docs for the new configuration

nazarhussain committed:
:white_check_mark: Add unit test for the configuration

nazarhussain committed:
:bookmark: Update change log file

nazarhussain committed:
:white_check_mark: Fix failing unit test

nazarhussain commented:
I had tested it with @aksdevac [provided code repo](https://github.com/ChainSafe/web3.js/issues/3891#issuecomment-773148060) by linking `web3-eth` and `web3-core-method` packages and it worked. For the wallet connect provider is that it wait for 5 seconds for `newBlockHeaders` event and since it does not return then it fallback to polling and fetch the tx receipt and return full-filled promise.

Spacesai1or commented on packages/web3-core-method/src/index.js:
```diff
@@ -56,6 +56,7 @@ var Method = function Method(options) {
     this.transactionBlockTimeout = options.transactionBlockTimeout || 50;
     this.transactionConfirmationBlocks = options.transactionConfirmationBlocks || 24;
     this.transactionPollingTimeout = options.transactionPollingTimeout || 750;
+    this.blockHeaderTimeout = options.blockHeaderTimeout || 5; // 10 seconds
```
```suggestion
    this.blockHeaderTimeout = options.blockHeaderTimeout || 5; // 5 seconds
```

Spacesai1or replied:
I'm just going to commit this since it's a comment change, and so I can provide my approval without it going stale

Commit:
Update comment

I'm just going to commit this since it's a comment change, and so I can provide my approval without it going stale

Spacesai1or reviewed (dismissed):
Thank you for updating the docs!

jdevcs commented on packages/web3-core-method/src/index.js:
```diff
@@ -56,6 +56,7 @@ var Method = function Method(options) {
     this.transactionBlockTimeout = options.transactionBlockTimeout || 50;
     this.transactionConfirmationBlocks = options.transactionConfirmationBlocks || 24;
     this.transactionPollingTimeout = options.transactionPollingTimeout || 750;
+    this.blockHeaderTimeout = options.blockHeaderTimeout || 5; // 5 seconds
```
Good to make `this.blockHeaderTimeout` default > or 10 sec as normal mining time is above that https://etherscan.io/chart/blocktime

jdevcs commented on packages/web3-core-method/src/index.js:
```diff
@@ -547,22 +548,35 @@ Method.prototype._confirmTransaction = function (defer, result, payload) {
-                }
-            });
+        // If provider do not support event subscription use polling
+        if(!this.requestManager.provider.on) {
```
Shouldnt we look specific event subscription support  in provider that we are interested in `newBlockHeaders` ,

nazarhussain replied:
There is no way to check if a provider a particular event or not.

jdevcs commented on packages/web3-eth/src/index.js:
```diff
@@ -100,6 +100,7 @@ var Eth = function Eth() {
     var transactionBlockTimeout = 50;
     var transactionConfirmationBlocks = 24;
     var transactionPollingTimeout = 750;
+    var blockHeaderTimeout = 5;
```
same here should be 10 sec

imaksp commented:
it seems many files are not using es6 & es7 features, is there any plan to update it.?

nazarhussain commented:
@aksdevac Yes for 1.x we are not refactoring the code a lot. As our focus is on the 4.x rewrite, which will cover best coding practices and Typescript language features.

nazarhussain committed:
:art: Update default value for blockHeaderTimeout to 10 seconds

xale76 commented:
please help. I don't understand how to let it work. No events received after send calls

nazarhussain commented:
@xale76 Hey, Can you please explain your comment. Are you referring to this PR? or to the issue itself? And how did you subscribed to those events which you mentioned not receiving?

xale76 commented:
@nazarhussain hello, and thanks. I'm using web3modal together with walletconnectprovider. My react app works perfectly from browser and metamask extension (connected by infura node on rinkeby). If I use walletconnectprovider, so I use QR code and I confirm actions on metamask app on my android mobile, after the confirmation (in example an "approve" action), my code remains stuck in "await". In example 

await mytoken.methods.approve(spender,amount).send({from:address});

I don't know what you mean with "subscribed" to events. I don't know, what events you're talking about, and I think it's not easy to list every events and register for them. Is there any other solution?

nazarhussain commented:
@xale76 This PR is about to fix that problem. Did you get the error even after compiling and linking this branch locally?

xale76 commented:
@nazarhussain how to install this branch from npm?

nazarhussain committed:
Merge branch '1.x' into nh/3891-receipt-callback

nazarhussain committed:
Merge branch '1.x' into nh/3891-receipt-callback

nazarhussain commented:
@xale76 You can not install the branch from npm. But this issue is scheduled for next release soon it will be available for you to test.

hoductrong commented:
Any update for this issues, or which release contains this fix?, plz

jdevcs commented:
@hoductrong in 1.5.3 https://github.com/ChainSafe/web3.js/pull/4304/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edR440
